File: extension/core/src/main/java/io/holunda/camunda/bpm/data/writer/CaseServiceVariableWriter.java
package io.holunda.camunda.bpm.data.writer;

import io.holunda.camunda.bpm.data.factory.VariableFactory;
import org.camunda.bpm.engine.CaseService;
import org.camunda.bpm.engine.variable.VariableMap;
import org.jetbrains.annotations.NotNull;

import java.util.Objects;

/**
 * Process execution builder allowing for fluent variable setting.
 */
public class CaseServiceVariableWriter implements VariableWriter<CaseServiceVariableWriter> {

  private final CaseService caseService;
  private final String caseExecutionId;

  /**
   * Creates a writer working on process execution.
   */
  public CaseServiceVariableWriter(CaseService caseService, String caseExecutionId) {
    this.caseService = caseService;
    this.caseExecutionId = caseExecutionId;
  }

  @Override
  @NotNull
  public VariableMap variables() {
    return this.caseService.getVariablesTyped(this.caseExecutionId);
  }

  @Override
  @NotNull
  public VariableMap variablesLocal() {
    return this.caseService.getVariablesLocalTyped(this.caseExecutionId);
  }

  @Override
  @NotNull
  public <T> CaseServiceVariableWriter set(VariableFactory<T> factory, T value) {
    return this.set(factory, value, false);
  }

  @Override
  @NotNull
  public <T> CaseServiceVariableWriter set(VariableFactory<T> factory, T value, boolean isTransient) {
    factory.on(this.caseService, this.caseExecutionId).set(value, isTransient);
    return this;
  }

  @Override
  @NotNull
  public <T> CaseServiceVariableWriter setLocal(VariableFactory<T> factory, T value) {
    return this.setLocal(factory, value, false);
  }

  @Override
  @NotNull
  public <T> CaseServiceVariableWriter setLocal(VariableFactory<T> factory, T value, boolean isTransient) {
    factory.on(this.caseService, this.caseExecutionId).setLocal(value, isTransient);
    return this;
  }

  @Override
  @NotNull
  public <T> CaseServiceVariableWriter remove(VariableFactory<T> factory) {
    factory.on(this.caseService, this.caseExecutionId).remove();
    return this;
  }

  @Override
  @NotNull
  public <T> CaseServiceVariableWriter removeLocal(VariableFactory<T> factory) {
    factory.on(this.caseService, this.caseExecutionId).removeLocal();
    return this;
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) return true;
    if (o == null || getClass() != o.getClass()) return false;

    CaseServiceVariableWriter that = (CaseServiceVariableWriter) o;

    if (!Objects.equals(caseService, that.caseService)) return false;
    return Objects.equals(caseExecutionId, that.caseExecutionId);
  }

  @Override
  public int hashCode() {
    int result = caseService != null ? caseService.hashCode() : 0;
    result = 31 * result + (caseExecutionId != null ? caseExecutionId.hashCode() : 0);
    return result;
  }
}

File: extension/core/src/main/kotlin/io/holunda/camunda/bpm/data/writer/CaseServiceVariableWriter.kt
package io.holunda.camunda.bpm.data.writer

import io.holunda.camunda.bpm.data.factory.VariableFactory
import org.camunda.bpm.engine.CaseService
import org.camunda.bpm.engine.variable.VariableMap
import java.util.function.Function

/**
 * Process execution builder allowing for fluent variable setting.
 */
class CaseServiceVariableWriter(private val caseService: CaseService, private val caseExecutionId: String) :
  VariableWriter<CaseServiceVariableWriter> {

  override fun variables(): VariableMap {
    return caseService.getVariablesTyped(caseExecutionId)
  }

  override fun variablesLocal(): VariableMap {
    return caseService.getVariablesLocalTyped(caseExecutionId)
  }

  override fun <T> set(variableFactory: VariableFactory<T>, value: T): CaseServiceVariableWriter {
    return this.set(variableFactory, value, false)
  }

  override fun <T> set(variableFactory: VariableFactory<T>, value: T, isTransient: Boolean): CaseServiceVariableWriter {
    variableFactory.on(caseService, caseExecutionId)[value] = isTransient
    return this
  }

  override fun <T> setLocal(variableFactory: VariableFactory<T>, value: T): CaseServiceVariableWriter {
    return this.setLocal(variableFactory, value, false)
  }

  override fun <T> setLocal(variableFactory: VariableFactory<T>, value: T, isTransient: Boolean): CaseServiceVariableWriter {
    variableFactory.on(caseService, caseExecutionId).setLocal(value, isTransient)
    return this
  }

  override fun <T> updateLocal(
    variableFactory: VariableFactory<T>,
    valueProcessor: Function<T, T>
  ): CaseServiceVariableWriter {
    return updateLocal(variableFactory, valueProcessor, false)
  }

  override fun <T> updateLocal(
    variableFactory: VariableFactory<T>,
    valueProcessor: Function<T, T>,
    isTransient: Boolean
  ): CaseServiceVariableWriter {
    variableFactory.on(caseService, caseExecutionId).updateLocal(valueProcessor, isTransient)
    return this
  }

  override fun <T> remove(variableFactory: VariableFactory<T>): CaseServiceVariableWriter {
    variableFactory.on(caseService, caseExecutionId).remove()
    return this
  }

  override fun <T> removeLocal(variableFactory: VariableFactory<T>): CaseServiceVariableWriter {
    variableFactory.on(caseService, caseExecutionId).removeLocal()
    return this
  }

  override fun <T> update(variableFactory: VariableFactory<T>, valueProcessor: Function<T, T>): CaseServiceVariableWriter {
    variableFactory.on(caseService, caseExecutionId).update(valueProcessor)
    return this
  }

  override fun <T> update(
    variableFactory: VariableFactory<T>,
    valueProcessor: Function<T, T>,
    isTransient: Boolean
  ): CaseServiceVariableWriter {
    variableFactory.on(caseService, caseExecutionId).updateLocal(valueProcessor, isTransient)
    return this
  }

  override fun equals(other: Any?): Boolean {
    if (this === other) return true
    if (other == null || javaClass != other.javaClass) return false
    val that = other as CaseServiceVariableWriter
    return if (caseService != that.caseService) false else caseExecutionId == that.caseExecutionId
  }

  override fun hashCode(): Int {
    var result = caseService.hashCode()
    result = 31 * result + caseExecutionId.hashCode()
    return result
  }
}
